File: leetcode/linked-lists/21-merge-two-sorted-lists/c/solution.c
#include <stdlib.h>

/// {{
/// Problem: 21. Merge Two Sorted Lists  
/// Difficulty: Easy  
/// Links: https://leetcode.com/problems/merge-two-sorted-lists/  
/// Topics: `linked-list`, `recursion`  
/// Timestamp: `Thu, 19 Jun 2025 18:30:55 +0530`

// Definition for singly-linked list.
struct ListNode {
  int val;
  struct ListNode *next;
};

struct ListNode *mergeTwoLists(
    struct ListNode *list1,
    struct ListNode *list2)
{
  struct ListNode dest = { .val = 0, .next = NULL };
  struct ListNode *cur1 = list1, *cur2 = list2, *cur3 = &dest;

  for (;;) {
    if (cur1 == NULL) {
      cur3->next = cur2;
      break;
    } else if (cur2 == NULL) {
      cur3->next = cur1;
      break;
    } else if (cur2->val < cur1->val) {
      cur3->next = cur2;
      cur2 = cur2->next;
      cur3 = cur3->next;
    } else {
      cur3->next = cur1;
      cur1 = cur1->next;
      cur3 = cur3->next;
    }
  }

  return dest.next;
}

/// }}

File: leetcode/linked-lists/21-merge-two-sorted-lists/ts/solution.ts
/// {{
/// Problem: 21. Merge Two Sorted Lists
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/merge-two-sorted-lists/
/// Topics: `linked-lsit`, `recursion`
/// Timestamp: `Wed, 17 Sep 2025 16:49:58 +0530`

// Definition for singly-linked list.
class ListNode {
  val: number
  next: ListNode | null
  constructor(val?: number, next?: ListNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }
}

function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null,
): ListNode | null {
  const l3 = new ListNode();
  let n1 = list1, n2 = list2, n3 = l3;

  for (; ;) {
    if (n1 === null) {
      n3.next = n2;
      break;
    } else if (n2 === null) {
      n3.next = n1;
      break;
    } else if (n2.val < n1.val) {
      n3.next = n2;
      n2 = n2.next;
      n3 = n3.next;
    } else {
      n3.next = n1;
      n1 = n1.next;
      n3 = n3.next;
    }
  }

  return l3.next;
}

/// }}
